pkg/security/admission/admission.go

```diff
@@ -0,0 +1,385 @@
+	"github.com/openshift/origin/pkg/security/uid"
+
+	"github.com/golang/glog"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+var _ kadmission.Interface = &constraint{}
+
+func NewConstraint(kclient client.Interface) kadmission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kclient.SecurityContextConstraints().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kclient.SecurityContextConstraints().Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&kapi.SecurityContextConstraints{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+// Admit determines if the pod should be admitted based on the requested security context
+// and the available SCCs.
+//
+// 1.  Find SCCs for the user.
+// 2.  Find SCCs for the SA.  If there is an error retrieving SA SCCs it is not fatal.
+// 3.  Remove duplicates between the user/SA SCCs.
+// 4.  Create the providers, includes setting pre-allocated values if necessary.
+// 5.  Try to generate and validate an SCC with providers.  If we find one then admit the pod
+//     with the validated SCC.  If we don't find any reject the pod and give all errors from the
+//     failed attempts.
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// get all constraints that are usable by the user
+	glog.V(4).Infof("getting security context constraints for pod %s (generate: %s) in namespace %s with user info %v", pod.Name, pod.GenerateName, a.GetNamespace(), a.GetUserInfo())
+	matchedConstraints, err := getMatchingSecurityContextConstraints(c.store, a.GetUserInfo())
+	if err != nil {
+		return kadmission.NewForbidden(a, err)
+	}
+
+	// get all constraints that are usable by the SA
+	saConstraints, err := c.getSecurityContextConstraintsForServiceAccount(a.GetNamespace(), pod)
+	if err != nil {
+		glog.Errorf("error getting security context constraints for service account %s: %v.  Only user constraints will be used.", pod.Spec.ServiceAccount, err)
+	} else {
+		matchedConstraints = append(matchedConstraints, saConstraints...)
+	}
+
+	// remove duplicate constraints and sort
+	matchedConstraints = deDupeSecurityContextConstraints(matchedConstraints)
```

Comment:
Suggest just calling this func `deduplicateSecurityContextConstraints`.  If you keep `dedup`, it should be in this case.